File: python/templates/outreach-service/app/outreach.py
"""Orchestration: the campaign-build-and-launch flow, composed from SDK calls.
Each function is a thin wrapper — the SDK does the API work; we add sequencing,
batching, and polling. Grounded in docs/api/* and workflows.json.
"""
from typing import Any, Dict, List, Optional

from instantly import create_instantly_client, InstantlyClient
from instantly.resources.leadlist import create_lead_list, get_verification_stats
from instantly.resources.lead import bulk_add_leads
from instantly.resources.supersearchenrichment import enrich_leads_from_supersearch
from instantly.resources.campaign import create_campaign, get_campaign_sending_status, activate_campaign
from instantly.resources.account import list_account
from instantly.resources.webhook import create_webhook


def get_client(api_key: Optional[str]) -> InstantlyClient:
    if not api_key:
        raise ValueError("INSTANTLY_API_KEY is not set")
    return create_instantly_client(api_key)


def create_list(client: InstantlyClient, name: str) -> str:
    result = create_lead_list(client, {"body": {"name": name}})
    return result.get("id", "") if isinstance(result, dict) else ""


def enrich_into_list(client: InstantlyClient, name: str, search_filters: Dict[str, Any], limit: int) -> str:
    run = enrich_leads_from_supersearch(client, {
        "body": {"search_filters": search_filters, "limit": limit, "list_name": name, "work_email_enrichment": True},
    })
    return run.get("resource_id", "") if isinstance(run, dict) else ""


def add_and_verify_leads(client: InstantlyClient, list_id: str, leads: List[Dict[str, Any]],
                         verify: bool = False) -> Any:
    # verify_leads_on_import spends a credit per lead — opt in explicitly.
    for i in range(0, len(leads), 1000):
        bulk_add_leads(client, {
            "body": {
                "list_id": list_id,
                "leads": leads[i:i + 1000],
                "verify_leads_on_import": verify,
                "skip_if_in_workspace": True,
            },
        })
    return get_verification_stats(client, {"path": {"id": list_id}})


def create_campaign_draft(client: InstantlyClient, params: Dict[str, Any]) -> str:
    campaign = create_campaign(client, {
        "body": {
            "name": params["name"],
            "campaign_schedule": {
                "schedules": [{
                    "name": "Business hours",
                    "timing": {"from": "09:00", "to": "17:00"},
                    "days": {"1": True, "2": True, "3": True, "4": True, "5": True},
                    "timezone": "America/Chicago",
                }],
            },
            "sequences": [{"steps": [{"type": "email", "delay": 0,
                                      "variants": [{"subject": params["subject"], "body": params["body"]}]}]}],
            "email_list": params.get("senders", []),
            "stop_on_reply": True,
        },
    })
    return campaign.get("id", "") if isinstance(campaign, dict) else ""


def preflight(client: InstantlyClient, campaign_id: str) -> Dict[str, Any]:
    status = get_campaign_sending_status(client, {"path": {"id": campaign_id}})
    accounts = list_account(client, {"query": {"limit": 100}})
    return {"status": status, "accounts": accounts}


def launch(client: InstantlyClient, campaign_id: str) -> Any:
    return activate_campaign(client, {"path": {"id": campaign_id}})


def register_reply_webhook(client: InstantlyClient, target_url: str, secret: str) -> Any:
    return create_webhook(client, {
        "body": {"target_hook_url": target_url, "event_type": "reply_received",
                 "headers": {"x-shared-secret": secret}},
    })

File: js/templates/outreach-service/src/outreach.ts
/**
 * Orchestration: the campaign-build-and-launch flow, composed from SDK calls.
 * Each function is a thin wrapper — the SDK does the API work; we add sequencing,
 * batching, and polling. Grounded in docs/api/* and workflows.json.
 */
import {
  createInstantlyClient,
  createLeadList,
  bulkAddLeads,
  enrichLeadsFromSupersearch,
  getVerificationStats,
  createCampaign,
  getCampaignSendingStatus,
  listAccount,
  activateCampaign,
  createWebhook,
  type InstantlyClient,
} from "@instantly-ai/sdk"

export function getClient(apiKey: string | undefined): InstantlyClient {
  if (!apiKey) throw new Error("INSTANTLY_API_KEY is not set")
  return createInstantlyClient({ apiKey })
}

/** Read an `id` off a loosely-typed success response. */
function idOf(value: unknown): string {
  return (value as { id?: string }).id ?? ""
}

export interface LeadInput {
  email: string
  first_name?: string
  last_name?: string
  company_name?: string
}

export interface BuildParams {
  name: string
  subject: string
  body: string
  senders: string[]
  leads?: LeadInput[]
  /** Verify leads on import — spends a credit per lead. Default off. */
  verify?: boolean
  /** Optional SuperSearch filters — if given, enrich into the list first. */
  searchFilters?: Record<string, unknown>
  enrichLimit?: number
}

/** Create a lead list; returns its id. */
export async function createList(client: InstantlyClient, name: string): Promise<string> {
  const list = await createLeadList(client, { body: { name } })
  return idOf(list)
}

/** Optionally source leads via SuperSearch into a new list; returns the resource/list id. */
export async function enrichIntoList(
  client: InstantlyClient,
  name: string,
  searchFilters: Record<string, unknown>,
  limit: number,
): Promise<string> {
  const run = await enrichLeadsFromSupersearch(client, {
    body: { search_filters: searchFilters as never, limit, list_name: name, work_email_enrichment: true },
  })
  return (run as { resource_id?: string }).resource_id ?? ""
}

/** Add BYO leads in batches of ≤1000, verifying on import. Returns list verification stats. */
export async function addAndVerifyLeads(
  client: InstantlyClient,
  listId: string,
  leads: LeadInput[],
  verify = false, // verify_leads_on_import spends a credit per lead — opt in explicitly
): Promise<unknown> {
  for (let i = 0; i < leads.length; i += 1000) {
    await bulkAddLeads(client, {
      body: {
        list_id: listId,
        leads: leads.slice(i, i + 1000),
        verify_leads_on_import: verify,
        skip_if_in_workspace: true,
      },
    })
  }
  return getVerificationStats(client, { path: { id: listId } })
}

/** Create the campaign as a Draft (status 0 — does NOT send). Returns campaign id. */
export async function createCampaignDraft(client: InstantlyClient, params: BuildParams): Promise<string> {
  const campaign = await createCampaign(client, {
    body: {
      name: params.name,
      campaign_schedule: {
        schedules: [
          {
            name: "Business hours",
            timing: { from: "09:00", to: "17:00" },
            days: { "1": true, "2": true, "3": true, "4": true, "5": true },
            timezone: "America/Chicago",
          },
        ],
      },
      sequences: [{ steps: [{ type: "email", delay: 0, variants: [{ subject: params.subject, body: params.body }] }] }],
      email_list: params.senders,
      stop_on_reply: true,
    },
  })
  return idOf(campaign)
}

/** Extract a list from a paginated response without assuming the exact envelope. */
function itemsOf(res: unknown): unknown[] | null {
  if (Array.isArray(res)) return res
  const r = res as Record<string, unknown> | null
  for (const k of ["items", "data", "accounts", "results"]) {
    if (r && Array.isArray(r[k])) return r[k] as unknown[]
  }
  return null
}

/**
 * Preflight: readiness check before activating. Never sends. Returns a summary
 * you can gate `launch()` on — connected-sender count, sending status, and
 * (when `listId` is given) verification stats. `hasSenders` is `false` only when
 * we definitively parsed zero senders, `"unknown"` if the payload shape was
 * unexpected — don't hard-block on `"unknown"`.
 */
export async function preflight(client: InstantlyClient, campaignId: string, listId?: string) {
  const status = await getCampaignSendingStatus(client, { path: { id: campaignId } })
  const accounts = await listAccount(client, { query: { limit: 100 } })
  const senders = itemsOf(accounts)
  const senderCount = senders ? senders.length : null
  const hasSenders: boolean | "unknown" = senderCount === null ? "unknown" : senderCount > 0
  // Verification is opt-in in this kit, so it does NOT gate readiness — surfaced for the caller to judge.
  const verification = listId ? await getVerificationStats(client, { path: { id: listId } }) : undefined
  const ready = hasSenders === true
  return { ready, checks: { hasSenders, senderCount }, status, accounts, verification }
}

/** Launch: activate the campaign (starts sending). Call only after preflight passes. */
export async function launch(client: InstantlyClient, campaignId: string): Promise<unknown> {
  return activateCampaign(client, { path: { id: campaignId } })
}

/** Register the reply webhook pointing at this service (custom-header auth; no HMAC). */
export async function registerReplyWebhook(
  client: InstantlyClient,
  targetUrl: string,
  secret: string,
): Promise<unknown> {
  return createWebhook(client, {
    body: { target_hook_url: targetUrl, event_type: "reply_received", headers: { "x-shared-secret": secret } },
  })
}
